File: algorithms/Binary Tree Postorder Traversal/Binary Tree Postorder Traversal.cpp
class Solution {
    void solve(TreeNode *root, vector<int> &ans){
        if(root == NULL) return;
        solve(root->left, ans);
        solve(root->right, ans);
        ans.push_back(root->val);
    }
public:
    vector<int> postorderTraversal(TreeNode* root) {
        vector<int> ans;
        solve(root, ans);
        return ans;
    }
};

File: scripts/algorithms/B/Binary Tree Postorder Traversal/Binary Tree Postorder Traversal.py
from typing import List, Optional


class Solution:
	"""
	Time:   O(n)
	Memory: O(n)
	"""

	def postorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
		if root is None:
			return []

		postorder = []
		stack = [root]

		while stack:
			node = stack.pop()
			postorder.append(node.val)
			if node.left is not None:
				stack.append(node.left)
			if node.right is not None:
				stack.append(node.right)

		return postorder[::-1]


class Solution:
	"""
	Time:   O(n)
	Memory: O(n)
	"""

	def postorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
		return list(self.postorder_generator(root))

	@classmethod
	def postorder_generator(cls, tree: Optional[TreeNode]):
		if tree is not None:
			yield from cls.postorder_generator(tree.left)
			yield from cls.postorder_generator(tree.right)
			yield tree.val